### PR TITLE
[RNMobile] Search Block: Implement long button text handling

### PR DESCRIPTION
**Pull request by AmandaRiu**
**Gutenberg Mobile PR:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3210

## Description

This PR implements the design for moving the button below the input element if the button text makes the button grow to more than 50% of the available width of the block:

Screen Shot 2021-04-14 at 9 28 06 PM

**Changes include the following:** 
- If the button is more the 50% in portrait mode, but not more than 50% in landscape mode, the view will appropriately reflect these changes when changing orientation. 
- If the button is moved below the search input element due to the length of text, and then the user changes to icon only button, the button will be moved back up to be inline with the input element. 
- Updated snapshots for unit tests

**Known issues/not included**
- Android text wraps before the text completely fills the button width. I'm not sure why this happens, but it doesn't appear to be related to my code directly (please let me know if I'm wrong about this). I've verified `maxWidth` is set appropriately but the text will still wrap before it fills the full width. 

## How has this been tested?

Test the following scenarios on iOS simulator and Android emulator. 

#### Long button text moves button below search input element
1. Add the Search Block. 
2. Click into the Search button text field and enter enough text to verify that the button is moved below the search input field when the width reaches about 50%. 
3. Delete the extra button text to verify the button will be moved inline with the search input field when expected. 

#### Long text button moves back inline with search input when icon only option enabled
1. Add the Search Block. 
2. Click into the Search button text field and enter enough text to verify that the button is moved below the search input field when the width reaches about 50%. 
3. Open the block settings.
4. Toggle icon only button to "on" and verify button is moved back up to be inline with search input element. 

#### Long button with button position set to "inside"
1. Add the Search Block. 
2. Click into the Search button text field and enter enough text to verify that the button is moved below the search input field when the width reaches about 50%. 
3. Open block settings.
4. Change "button position" option to "Inside". Verify design. 

#### Long button recalculated on orientation change
1. Add the Search Block in portrait mode.
2. Click into the Search button text field and enter _just enough text_ so the button moves below the search input element. 
3. Rotate the device to landscape mode. Verify the button moves back up to be inline with the search input element. 
4. Enter more text in the button to verify that once the button is about 50% of the width it gets moved below the search input element. 

#### Interacting with the UI

https://user-images.githubusercontent.com/5810477/114802279-6e265700-9d6b-11eb-815e-e23e49709705.mov

#### Rotating on Android

https://user-images.githubusercontent.com/5810477/114802233-5b138700-9d6b-11eb-97e3-c0d19ebe61c2.mov

#### Rotating on iOS

https://user-images.githubusercontent.com/5810477/114802266-6961a300-9d6b-11eb-928f-2ffa820b8b83.mov

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by AmandaRiu** [comment]
cc: @iamthomasbishop

File: packages/block-library/src/search/edit.native.js
Comment (geriux) [inline packages/block-library/src/search/edit.native.js]: I think here we could use the prop `blockWidth` that every block receives in the `props` from the parent wrapper. This way we can only have just one `onLayout` for the button. What do you think?

You have it available [here](https://github.com/WordPress/gutenberg/blob/e4a5461b98b6cb9da62f8a4add4e42a3cce59a09/packages/block-library/src/search/edit.native.js#L45).

Comment (geriux) [reply packages/block-library/src/search/edit.native.js]: Something related to `Dimensions` to take into account is that depending on where/how we use it we might need to add a listener to dimensions changes, an example: iPad in split view, if you resize the space you might end up with the first value you got from the Dimensions helper.

I _think_ you won't need this if you use the `blockWidth` prop because it will trigger a re-render from the parent [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/readable-content-view/index.native.js).

There's `useWindowDimensions` from React native but there's a bug in the version we have now, once we update we could use that 🤞

Comment (AmandaRiu) [reply packages/block-library/src/search/edit.native.js]: Awesome, thanks for the tips @geriux , I was wondering if there was a way to get the `blockWidth` so this is really helpful.

Comment (AmandaRiu) [reply packages/block-library/src/search/edit.native.js]: Updated in 0fcfd43fe01f3edfcd12cd3694950fdbf820ffcd

**Comment by AmandaRiu** [comment]
@geriux I've made the suggested changes. Thank you for the tip, that really helped to remove a lot of unnecessary code.

**Review by iamthomasbishop — COMMENTED** [review]
This looks really great, @AmandaRiu. I definitely prefer this approach over the web's because it more gracefully accommodates long labels, but I worry that despite the layout looking like a cluster, folks will be a little confused when they see a mismatch between the layout in the mobile editor vs. the front-end output. I think this is one of those awkward cases where we have to decide whether we should: 

1. Align with the front-end output (and core) and have a...less-than-ideal UI in the editor in order to match with the output, or...
2. Make it look nicer in the editor (like you have here) which will result in a diff between what the user is seeing in the editor and on the front-end of their site

As much as it pains me to say, I think it'd be better to align with styling of the output (at least in terms of layout) for the time being. However, I believe this approach would be beneficial everywhere, so we should propose it upstream to core Gutenberg. I understand this isn't ideal, but it might be worth considering for a long-term solution.

@kyleaparker what do you think? Is it worth diverging from the output in this case, wrt layout behavior?

---

Note: my assumption here is that the front-end result that we're producing uses the same exact markup that is applied using the web editor — if I'm wrong and we're applying our own set of styles to make the block behave as it does in the mobile editor, then by all means, let's roll with this 😃

**Comment by AmandaRiu** [comment]
> Note: my assumption here is that the front-end result that we're producing uses the same exact markup that is applied using the web editor — if I'm wrong and we're applying our own set of styles to make the block behave as it does in the mobile editor, then by all means, let's roll with this 😃

Hi @iamthomasbishop 👋  This design is specific and custom to the mobile editor and was in the i2 designs for the mobile version of the search block. I believe @kyleaparker approached this much like we approach the columns block in the mobile editor which shows each column as a row if space is limited. Editing the columns on mobile will look differently than web which may have some drawbacks, but the experience and UX is so much better than if we tried to make everything on mobile look just like it would on the web. 

The Search block is due at the end of this sprint (tomorrow) and this PR is blocking the final PR that will put the Search block into production so I'd like to get a final yay or nay on this functionality if possible. Thank you for the review!

**Comment by kyleaparker** [comment]
> @kyleaparker what do you think? Is it worth diverging from the output in this case, wrt layout behavior?

I think it is! If you're editing your site on mobile we should make sure you have the most optimal user experience. We should definitely push for this change for the web because long button text alongside a search input looks a little broken right now. We should also push towards a front end that matches and provides the best mobile experience for site visitors. 

FWIW, I don't think too many people will deviate from the Search text so I imagine the number of people who see this layout will be small.

**Comment by iamthomasbishop** [comment]
@AmandaRiu To be clear, I don't think this is worth blocking the rest of the project — I would consider it somewhat of an edge case that will only be experienced if you have a super-long text label. With that said, I did want to discuss because there is a potential for a relatively large mismatch between what the user sees on the front-end vs. in the editor. Definitely not worth holding up the entire project for, but rather something we can decide and iterate on.

I get where you're coming from and 100% agree that the interaction model here is better from a UX perspective inside the editor, but if the produced layout on the front-end is substantially different (in this case, worse) then it's bound to cause confusion (in this particular case where they have a very long label — again, probably an edge case). 

Regarding Columns, unless there's a  new bug I'm not aware of, layouts that you build in the mobile editor should be laid out on the front-end in roughly the same way and utilize the same breakpoints that are provided by core GB. IIRC, when we worked on the initial versions of Columns we made the same sort of decision — to follow the same breakpoints and layout that exist on the web (for better or worse).

**Comment by iamthomasbishop** [comment]
@kyleaparker I agree w/ most of what you're saying; it does feel counter-intuitive to suggest that we _not_ default to the absolute best authoring experience. 😄 I just wanted to voice my concern about the potential frustration a user might experience when they've added a long label in the editor (which adapts nicely as shown above), but sees that the result on the front-end is janky — but that's part of a much bigger conversation. Considering the scale of impact for better or worse is small in this case, I'm totally fine with whatever y'all think is best here!

**Comment by AmandaRiu** [comment]
@kyleaparker @iamthomasbishop I appreciate the open discussion and thank you for all your input. I agree there is so much gray area here and it's a difficult and very important thing to be aware of so we can create the best possible experience for the user. That said, it sounds like we are keeping these changes so this PR is good to merge once approved. 👍

**Comment by AmandaRiu** [comment]
NOTE: [Opened a ticket](https://github.com/WordPress/gutenberg/issues/30884) to have this same functionality ported to the mobile web version as well.

File: packages/block-library/src/search/edit.native.js
Comment (dcalhoun) [inline packages/block-library/src/search/edit.native.js:301]: I did notice one oddity, but I'm not sure what portion of the code to attribute it to. When entering text into the button that causes the button to expand its width, the button resizes to fit the new character a split second _after_ the text is appended. There is a brief moment where the new character is visible within the "padding" of the button. It's not terribly disruptive, but creates a small bit of jumping in the layout. Likely not worth attempting to address in this PR, but I wanted to point it out nonetheless.

https://user-images.githubusercontent.com/438664/114952827-c0738080-9e1c-11eb-9293-5956db70f2e1.mov

**Review by dcalhoun — APPROVED** [review]
LGTM. I tested this an iPhone SE and the Android emulator. Operated as expected, aside from one comment regarding the button width expanding while typing.

File: packages/block-library/src/search/edit.native.js
Comment (dcalhoun) [reply packages/block-library/src/search/edit.native.js:301]: After looking at this one last time, I'm thinking this oddity is merely the nature of how the `RichText` and Aztec editor is implemented. I noticed the Button block also suffers from this same behavior, so it is not related to the styling of the Search block whatsoever.

**Comment by dcalhoun** [comment]
@AmandaRiu I did just run into an issue that may be worth investigating further. 😞 When (1) inserting the Search block while having (2) a text block focused _and_  (3) using the block inserter search, an infinite focus loop is created. This does not occur if I do not have all three of those criteria met. 

1. Launch the block editor. 
2. Tap the title of the editor to focus the text field. 
3. Tap the + button to open the block inserter. 
4. Search for the Search block. 
5. Select the search block.

An infinite focus loop is created, which eventually would likely lead to a crash. 

This is not the first time I have seen an infinite loop. There are a few open issues — wordpress-mobile/gutenberg-mobile#953, wordpress-mobile/gutenberg-mobile#1696. However, this one was pretty easy for me to recreate. I have not yet tested this within the WPiOS and WPAndroid apps themselves, only the Demo editor app. I'm not sure if the issue would be better or worse in the actual apps. I'll try to find time help investigate, but it may be a bit before I can spend more time. 

https://user-images.githubusercontent.com/438664/114953700-8c00c400-9e1e-11eb-9f54-a4085608e627.MP4

**Comment by AmandaRiu** [comment]
> @AmandaRiu I did just run into an issue that may be worth investigating further. 😞 When (1) inserting the Search block while having (2) a text block focused _and_ (3) using the block inserter search, an infinite focus loop is created. This does not occur if I do not have all three of those criteria met.
> 
>     1. Launch the block editor.
> 
>     2. Tap the title of the editor to focus the text field.
> 
>     3. Tap the + button to open the block inserter.
> 
>     4. Search for the Search block.
> 
>     5. Select the search block.
> 
> 
> An infinite focus loop is created, which eventually would likely lead to a crash.
> 
> This is not the first time I have seen an infinite loop. There are a few open issues — [wordpress-mobile/gutenberg-mobile#953](https://github.com/wordpress-mobile/gutenberg-mobile/issues/953), [wordpress-mobile/gutenberg-mobile#1696](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1696). However, this one was pretty easy for me to recreate. I have not yet tested this within the WPiOS and WPAndroid apps themselves, only the Demo editor app. I'm not sure if the issue would be better or worse in the actual apps. I'll try to find time help investigate, but it may be a bit before I can spend more time.
> RPReplay_Final1618531729.MP4

Thanks for the review @dcalhoun! I tried to replicate the focus issue but am not having any luck. I tried on an Android emulator, as well as my Pixel 4 running Android 11. What are the specs of the device you were using for testing? 


https://user-images.githubusercontent.com/5810477/114956161-5236ba00-9e2c-11eb-9be4-b94aa65bf70d.mp4

**Comment by dcalhoun** [comment]
> What are the specs of the device you were using for testing? 

iPhone SE, iOS 14.4.2. Local Metro server running on this branch. I'll try to spend some time tomorrow investigating further.

**Comment by geriux** [comment]
Regarding the loop problem, I created an issue https://github.com/WordPress/gutenberg/issues/30562 for that (the search functionality is still in dev)

**Comment by AmandaRiu** [comment]
> > What are the specs of the device you were using for testing?
> 
> iPhone SE, iOS 14.4.2. Local Metro server running on this branch. I'll try to spend some time tomorrow investigating further.

Aha! Okay, I tested on my physical iPhone 11 running iOS v14.4.2 and can replicate this issue. It's not the search block however, it happens on any text block. Here's me testing with adding a paragraph block: 


https://user-images.githubusercontent.com/5810477/115280239-b64fcb80-a115-11eb-8385-f1df806b773a.mp4

**Pull request merged**
